pkg/claims/service.go

```diff
@@ -46,10 +49,163 @@ func NewService(treeStore db.Storage, signedClaimStore *claimsstore.SignedClaimP
+	}
+	// we get the claim with that index
+	rootClaimFromLookUp, err := s.rootClaimFromLookUp(rootClaim.Entry().HIndex(), rootSnapShotTree)
+	if err != nil {
+		return nil, errors.Wrap(err, "getLastRootClaim.rootClaimFromLookUp")
+	}
+
+	// get the next version which is the version after the last existing version
+	nextVersion, err := isrv.GetNextVersion(rootSnapShotTree, rootClaimFromLookUp.Entry().HIndex())
+	if err != nil {
+		return nil, errors.Wrap(err, "getLastRootClaim.GetNextVersion")
+	}
+	// set the version to the last existing version
+	rootClaimFromLookUp.Version = nextVersion - 1
+
+	// next we get the claim with this version
+	return s.rootClaimFromLookUp(rootClaimFromLookUp.Entry().HIndex(), rootSnapShotTree)
+}
+
+func (s *Service) getRootSnapshot(commit *claimsstore.RootCommit) (*merkletree.MerkleTree, error) {
+	lastRoot, err := hex.DecodeString(commit.Root[2:])
+	if err != nil {
+		return nil, errors.Wrap(err, "getRootSnapshot hex.DecodeString")
+	}
+	if len(lastRoot) != 32 {
+		return nil, errors.New("root hash should be 32 bytes")
+	}
+
+	lastrootHash := merkletree.Hash{}
+	copy(lastrootHash[:], lastRoot)
+
+	return s.rootMt.Snapshot(&lastrootHash)
+}
+
+func (s *Service) makeContentClaimFromCred(claim *claimsstore.ContentCredential) (*ClaimRegisteredDocument, error) {
+	signerDID, err := didlib.Parse(claim.Proof.Creator)
+	if err != nil {
+		return nil, errors.Wrap(err, "makeContentClaimFromCred didlib.Parse")
+	}
+	claimJSON, err := json.Marshal(claim)
+	if err != nil {
+		return nil, errors.Wrap(err, "makeContentClaimFromCred json.Marshal")
+	}
+	hash := crypto.Keccak256(claimJSON)
+	hash32 := [32]byte{}
+	copy(hash32[:], hash)
+	return NewClaimRegisteredDocument(hash32, signerDID, ContentCredentialDocType)
+}
+
+// GenerateProof returns a proof that the content credential is in the tree and on the blockchain
+func (s *Service) GenerateProof(claim *claimsstore.ContentCredential) (*MTProof, error) {
+	lastRootCommit, err := s.rootService.GetLatest()
+	if err != nil {
+		return nil, errors.Wrap(err, "GenerateProof.rootService.GetLatest")
+	}
+
+	lastRootSnapshot, err := s.getRootSnapshot(lastRootCommit)
+	if err != nil {
+		return nil, errors.Wrap(err, "GenerateProof.getRootSnapshot")
+	}
+
+	signerDID, err := didlib.Parse(claim.Proof.Creator)
+	if err != nil {
+		return nil, errors.Wrap(err, "GenerateProof didlib.Parse")
+	}
+
+	didMT, err := s.buildDIDMt(signerDID)
+	if err != nil {
+		return nil, errors.Wrap(err, "GenerateProof.buildDIDMt")
+	}
+
+	latestRootClaim, err := s.getLastRootClaim(didMT, lastRootSnapshot, signerDID)
+	if err != nil {
+		return nil, errors.Wrap(err, "GenerateProof.getLastRootClaim")
+	}
+
+	didRootExistsProof, err := lastRootSnapshot.GenerateProof(latestRootClaim.Entry().HIndex(), lastRootSnapshot.RootKey())
+	if err != nil {
+		return nil, errors.Wrap(err, "GenerateProof.lastRootSnapshot.GenerateProof")
+	}
+
+	didTreeSnapshot, err := didMT.Snapshot(&latestRootClaim.RootKey)
```

Comment:
So for my clarity, everything up to this point is to generate the tree snapshot correct?  Then the proof that the given claim is in the tree is produced in `generateProofAndNonRevokeEntry`?  If so, could potentially roll up into a method.

Reply:
Not everything, we also need to generate the proof that the root claim is in the root tree, but could probably be condensed a bit more